pkg/generators/forwarding/fluentd/fluent_conf_test.go

```diff
@@ -482,31 +503,69 @@ var _ = Describe("Generating fluentd config", func() {
 
   # Relabel specific sources (e.g. logs.apps) to multiple pipelines
   <label @_APPLICATION>
+     <match kubernetes.**_dev-apple_**>
+      @type copy
+      <store>
+	    @type relabel
+	    @label @APPS_PIPELINE2
+      </store>
+    </match>
+    <match kubernetes.**_project1-namespace_**>
+      @type copy
+      <store>
+		@type relabel
+		@label @APPS_PIPELINE
+      </store>
+    </match>
+    <match kubernetes.**_project2-namespace_**>
+      @type copy
+      <store>
+		@type relabel
+		@label @APPS_PIPELINE
+      </store>
+      <store>
+		@type relabel
+		@label @APPS_PIPELINE2
+      </store>
+    </match>
     <match **>
       @type copy
-
       <store>
-        @type relabel
-        @label @APPS_PIPELINE
+	    @type relabel
+	    @label @MY_DEFAULT_PIPELINE
```

Comment:
does this `<match>` order against namespace tag means `MY_DEFAULT_PIPELINE` receives all namespaces **except** `project1-namespace, project2-namespace, dev-apple` namespace? is that the intention?

https://docs.fluentd.org/configuration/config-file#note-on-match-order

Do we ensure a wider match appears lower in the order?

Reply:
I built forwarder generator from this PR and generated fluent.conf from below yaml. I think this yaml was part of a BZ about NS forwarding
```
apiVersion: logging.openshift.io/v1
kind: ClusterLogForwarder
metadata:
  name: instance
  namespace: openshift-logging
spec:
  inputs:
    - application:
        namespaces:
         - test
      name: specific-app-logs
  outputs:
   - name: rsyslog-ex
     type: syslog
     syslog:
       facility: local0
       rfc: RFC3164
       payloadKey: message
       severity: informational
     url: 'tcp://10.10.93.88:514'
  pipelines:
    - inputrefs:
        - application
        - audit
        - infrastructure
      name: application-audit-infrastructure-logs
      outputrefs:
        - default
    - inputrefs:
        - specific-app-logs
      name: forward-to-external
      outputrefs:
        - rsyslog-ex
```
It generates cofig with wider match on top. So perhaps we cannot guarantee match order in generated conf.


```
# Relabel specific sources (e.g. logs.apps) to multiple pipelines
<label @_APPLICATION>
  <match **> 
    @type copy
    <store>
      @type relabel
      @label @APPLICATION_AUDIT_INFRASTRUCTURE_LOGS
    </store>
  </match>
  <match kubernetes.**_test_**>      #<<<< this will never match
    @type copy
    <store>
      @type relabel
      @label @FORWARD_TO_EXTERNAL
    </store>
  </match>
</label>
<label @_AUDIT>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @APPLICATION_AUDIT_INFRASTRUCTURE_LOGS
    </store>
  </match>
</label>
<label @_INFRASTRUCTURE>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @APPLICATION_AUDIT_INFRASTRUCTURE_LOGS
    </store>
  </match>
</label>
```

Reply:
I will review your findings.  Each match block should be 1:1 to a namespace since we know where the namepace falls within the tag.  The blocks should be ordered such that ns matches occur before a catch all which is `**`.  If it does not do this then that is a bug.

Reply:
I see the issue.  This is because we have the concept of `application` which means all, and a subset of application logs which come from a defined pipeline.  It has to know how to deal with both

Reply:
actually, the test case looks like it covers this scenario, so need to investigate

Reply:
I think you need to add MY_DEFAULT_ROUTE to every copy stanza, because it receives from every namespace. Then the ** catch-all will add any namespaces that were not explicitly matched.

Is this really easier or more efficient than using https://github.com/banzaicloud/fluent-plugin-label-router? It does exactly what we want - define a set of routes, guarded by a set of label/namepsace matching rules.

Reply:
I have no familiarity with this plugin.  I was trying to solve a bz that has been outstanding since essentially GA.  I'm not sure what you are recommended here.  The only issue I can see was the "anything left over" should be caught by "**" has to go last, which I resolved.  The pipeline via labels is basically:  INGRESS->_TYPE ->PIPELINE->OUTPUT where application logs entering _TYPE are specifically copied to multiple PIPELINES as defined by the input namespacing.

Reply:
This may be off-topic, but please let me explain what I tried to do in https://github.com/openshift/cluster-logging-operator/pull/865.

As @alanconway said, label router plugin would help us to maintain a set of label/namepsace matching rules.

In https://github.com/openshift/cluster-logging-operator/pull/865, I tried to introduce Input Selector Block, which defines a set of routes by using label router plugin.
This block will re-label a subset of application logs based on namespaces as well as their Kubernetes metadata.
We will be able to centrally control  label/namespace matching rules by label router plugin. 
### Example LF CR:
```
spec:
  inputs:
   - name: app-product-logs
     application:
       namespaces:
         - test1
       selector
         matchLabels: 
           environment: product
   - name: app-dev-logs
     application:
       namespaces:
         - test2
       selector
         matchLabels: 
           environment: dev
  outputs:
   - name: syslog1
     ...
   - name: syslog2
     ...
  pipelines:
   - name: forward-product-logs-to-syslog1
     inputRefs:
     - app-product-logs
     outputRefs:
     - syslog1
   - name: forward-dev-logs-to-syslog2
     inputRefs:
     - app-dev-logs
     outputRefs:
     - syslog2
   - name: all-default
     inputRefs:
     - audit
     - infrastructure
     - application
     outputRefs:
     - default
```
### fluent.conf after applying Example LF CR:
```
<label @_APPLICATION>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
    <store>
      @type relabel
      @label @_APPLICATION_BASED_ON_INPUT_SELECTOR
    </store>
  </match>
</label>
<label @_APPLICATION_BASED_ON_INPUT_SELECTOR>
  <match **>
    @type label_router       # this will re-label application logs as defined by the input namespacing instead of copying to multiple PIPELINES
    <route>
      @label @FORWARD_PRODUCT_LOGS_TO_SYSLOG1
      <match>
        labels environment:product
        namespaces test1
      </match>
    </route>
    <route>
      @label @FORWARD_DEV_LOGS_TO_SYSLOG2
      <match>
        labels environment:dev
        namespaces test2
      </match>
    </route>
  </match>
</label>

<label @_AUDIT>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
  </match>
</label>

<label @_INFRASTRUCTURE>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
  </match>
</label>
```